Add migration guide (#352)
bakdata/kpops: docs/docs/user/migration-guide/v1-v2.md

# Migrate from V1 to V2

## [Derive component type automatically from class name](https://github.com/bakdata/kpops/pull/309)

> KPOps automatically inferes the `type` from the class name. Therefore, there is no need to add the `type` anymore. By convention the `type` would be the lower, and kebab cased name of the class.

Reviewer: ```suggestion
KPOps automatically infers the component `type` from the class name. Therefore, the `type` and `schema_type` attributes can be removed from your custom components. By convention the `type` would be the lower, and kebab cased name of the class.
```
TIL about the name kebab case 😂 I believe I called it dash case in the function

perhaps include the removal of `schema_type` in the diff as well

Author: I would like to stick with the kebab case 😄

```diff
class MyCoolStreamApp(StreamsApp):
-    type = "my-cool-stream-app"
+    ...
```

> The [ProducerApp](https://github.com/bakdata/kpops/blob/main/kpops/components/streams_bootstrap/producer/producer_app.py#L16) class, which had the `type` `producer`, now through these changes and name convention hast the `type` `producer-app`. This must be addressed in your `pipeline.yaml` or `defaults.yaml`.

Reviewer: use release tag instead of main

> The [ProducerApp](https://github.com/bakdata/kpops/blob/main/kpops/components/streams_bootstrap/producer/producer_app.py#L16) class, which had the `type` `producer`, now through these changes and name convention hast the `type` `producer-app`. This must be addressed in your `pipeline.yaml` or `defaults.yaml`.

Reviewer: imo just saying: "because of this new convention `producer` has been renamed to `producer-app`" would be enough

In the `from` section these have changed:

- The Default type is `input`
- `input-pattern` type is replaced by `pattern`
- If `role` is set, type is inferred to be `extra`
- If `role` is set, type is explicitly set to `pattern`, this would be inferred type `extra-pattern`

```diff
  from:
    topics:
      ${pipeline_name}-input-topic:
-       type: input
        ...
      ${pipeline_name}-extra-topic:
-       type: extra
        role: topic-role
        ...
      ${pipeline_name}-input-pattern-topic:
-       type: input-pattern 
+       type: pattern
        ...
      ${pipeline_name}-extra-pattern-topic:
-       type: extra-pattern 
+       type: pattern
        role: some-role
        ...
```

## [Remove camel case conversion of internal models](https://github.com/bakdata/kpops/pull/308)

All the internal KPOps models are now snake_case, and only Helm/Kubernetes values require camel casing. You can find an example of a `pipeline.yaml` in the following. Notice that the `app` section here remains untouched.

```diff
...
type: streams-app
  name: streams-app
  namespace: namespace 
  app: 
    streams:
      brokers: ${brokers} 
      schemaRegistryUrl: ${schema_registry_url}

>      autoscaling:d

Reviewer: ```suggestion
     autoscaling:
```

# Migrate from V1 to V2

## [Derive component type automatically from class name](https://github.com/bakdata/kpops/pull/309)

KPOps automatically infers the component `type` from the class name. Therefore, the `type` and `schema_type` attributes can be removed from your custom components. By convention the `type` would be the lower, and kebab cased name of the class.

```diff
class MyCoolStreamApp(StreamsApp):
-    type = "my-cool-stream-app"
+    ...
```

> Because of this new convention `producer` has been renamed to `producer-app`. This must be addressed in your `pipeline.yaml` or `defaults.yaml`.

Reviewer: ```suggestion
Because of this new convention `producer` has been renamed to `producer-app`. This must be addressed in your `pipeline.yaml` and `defaults.yaml`.
```

```diff
- producer:
+ producer-app:
    app:
        streams:
        outputTopic: output_topic
        extraOutputTopics: 
            output_role1: output_topic1
            output_role2: output_topic2
```

## [Refactor input/output types](https://github.com/bakdata/kpops/pull/232)

### To section

In the `to` section these have changed:

- The default type is `output`
- If `role` is set, type is inferred to be `extra`
- The type `error` needs to be defined explicitly

```diff
  to:
    topics:
      ${pipeline_name}-topic-1:
-       type: extra
        role: "role-1"
        ...
      ${pipeline_name}-topic-2:
-       type: output 
        ...
      ${pipeline_name}-topic-3:
         type: error 
         ...
```

### From section

In the `from` section these have changed:

> - The Default type is `input`

Reviewer: ```suggestion
- The default type is `input`
```
